### ISSUE TITLE
Transient "Local: Broker transport failure"

### ISSUE THREAD

**Issue by dshivashankar1994** [edenhill/kafkacat#238]
I have a thin wrapper over the librd Kafka Consumer which has a separate poll thread that would call the consume method. When the application does forking, the poll thread is paused and once the forking is done, the poll thread resumes.

At time once the thread resumes, I see the failure - "Local: Broker transport failure". I am unable to find a reproducer for this too.

**Comment by edenhill** [comment]
An rd_kafka_t instance may not be used across forks, it needs to be created in the same process that it will be used since threads will not survive a fork.

**Comment by dshivashankar1994** [comment]
Yes. I understand. But my application has capability to fork. So what I am doing is, I pause the thread which is consuming (ie poll thread) in case of forking and then resume the same thread after the main application forking is done. The poll thread continues to run in the same process only.

The issue I reported doesn't happen with every fork in the mail application. But only at times.

Do you think anything else had gone wrong here ?

**Comment by edenhill** [comment]
You can troubleshoot this further by setting the `debug` config property to `broker,cgrp`

**Comment by dshivashankar1994** [comment]
I ran my KafkaReader in debug mode and in logs, I could find the error messages

```
[127201 (140267268077312) - DEBUG 20200522 11:12:37.180 EDT KafkaReader.C#event_cb:72] 7, facility BROKERFAIL, str [thrd:GroupCoordinator]: GroupCoordinator: failed: err: Local: Broker transport failure: (errno: Connection reset by peer)
[127201 (140267268077312) - DEBUG 20200522 11:12:37.180 EDT KafkaReader.C#event_cb:72] 7, facility FAIL, str [thrd:GroupCoordinator]: GroupCoordinator: Disconnected (after 600029ms in state UP)
[127201 (140267268077312) - DEBUG 20200522 11:12:37.180 EDT KafkaReader.C#event_cb:72] 7, facility STATE, str [thrd:GroupCoordinator]: GroupCoordinator: Broker changed state UP -> DOWN
[127201 (140267268077312) - DEBUG 20200522 11:12:37.180 EDT KafkaReader.C#event_cb:72] 7, facility BROADCAST, str [thrd:GroupCoordinator]: Broadcasting state change
[127201 (140267268077312) - DEBUG 20200522 11:12:37.180 EDT KafkaReader.C#event_cb:72] 7, facility METADATA, str [thrd:GroupCoordinator]: Requesting metadata for 19/19 topics: broker down
[127201 (140267268077312) - DEBUG 20200522 11:12:37.180 EDT KafkaReader.C#event_cb:72] 7, facility METADATA, str [thrd:GroupCoordinator]: sasl_plaintext://mwkafka-staging-01.dr.myFirm.com:9092/3: Request metadata for 19 topic(s): broker down
[127201 (140267268077312) - DEBUG 20200522 11:12:37.180 EDT KafkaReader.C#event_cb:72] 7, facility STATE, str [thrd:GroupCoordinator]: GroupCoordinator: Broker changed state DOWN -> INIT
[127201 (140267268077312) - DEBUG 20200522 11:12:37.180 EDT KafkaReader.C#event_cb:72] 7, facility BROADCAST, str [thrd:GroupCoordinator]: Broadcasting state change
[127201 (140267268077312) - DEBUG 20200522 11:12:37.180 EDT KafkaReader.C#event_cb:72] 7, facility STATE, str [thrd:GroupCoordinator]: GroupCoordinator: Broker changed state INIT -> TRY_CONNECT
[127201 (140267268077312) - DEBUG 20200522 11:12:37.180 EDT KafkaReader.C#event_cb:72] 7, facility BROADCAST, str [thrd:GroupCoordinator]: Broadcasting state change
[127201 (140267268077312) - DEBUG 20200522 11:12:37.180 EDT KafkaReader.C#event_cb:72] 7, facility CONNECT, str [thrd:GroupCoordinator]: GroupCoordinator/1: broker in state TRY_CONNECT connecting
[127201 (140267268077312) - DEBUG 20200522 11:12:37.180 EDT KafkaReader.C#event_cb:72] 7, facility STATE, str [thrd:GroupCoordinator]: GroupCoordinator: Broker changed state TRY_CONNECT -> CONNECT
[127201 (140267268077312) - DEBUG 20200522 11:12:37.180 EDT KafkaReader.C#event_cb:72] 7, facility BROADCAST, str [thrd:GroupCoordinator]: Broadcasting state change
[127201 (140267268077312) - DEBUG 20200522 11:12:37.182 EDT KafkaReader.C#event_cb:72] 7, facility CONNECT, str [thrd:GroupCoordinator]: GroupCoordinator/1: Connecting to ipv4#10.219.247.10:9092 (sasl_plaintext) with socket 37
[127201 (140267268077312) - DEBUG 20200522 11:12:37.180 EDT KafkaReader.C#event_cb:72] 7, facility BROADCAST, str [thrd:GroupCoordinator]: Broadcasting state change
[127201 (140267268077312) - DEBUG 20200522 11:12:37.182 EDT KafkaReader.C#event_cb:72] 7, facility CONNECT, str [thrd:GroupCoordinator]: GroupCoordinator/1: Connecting to ipv4#10.219.247.10:9092 (sasl_plaintext) with socket 37
[127201 (140267268077312) - DEBUG 20200522 11:12:37.182 EDT KafkaReader.C#event_cb:72] 7, facility CONNECT, str [thrd:GroupCoordinator]: GroupCoordinator/1: Connected to ipv4#10.219.247.10:9092
[127201 (140267268077312) - DEBUG 20200522 11:12:37.182 EDT KafkaReader.C#event_cb:72] 7, facility CONNECTED, str [thrd:GroupCoordinator]: GroupCoordinator/1: Connected (#9)
[127201 (140267268077312) - DEBUG 20200522 11:12:37.182 EDT KafkaReader.C#event_cb:72] 7, facility STATE, str [thrd:GroupCoordinator]: GroupCoordinator: Broker changed state CONNECT -> APIVERSION_QUERY
[127201 (140267268077312) - DEBUG 20200522 11:12:37.182 EDT KafkaReader.C#event_cb:72] 7, facility BROADCAST, str [thrd:GroupCoordinator]: Broadcasting state change
[127201 (140267268077312) - DEBUG 20200522 11:12:37.183 EDT KafkaReader.C#event_cb:72] 7, facility AUTH, str [thrd:GroupCoordinator]: GroupCoordinator/1: Auth in state APIVERSION_QUERY (handshake supported)
[127201 (140267268077312) - DEBUG 20200522 11:12:37.183 EDT KafkaReader.C#event_cb:72] 7, facility STATE, str [thrd:GroupCoordinator]: GroupCoordinator: Broker changed state APIVERSION_QUERY -> AUTH_HANDSHAKE
[127201 (140267268077312) - DEBUG 20200522 11:12:37.183 EDT KafkaReader.C#event_cb:72] 7, facility BROADCAST, str [thrd:GroupCoordinator]: Broadcasting state change
[127201 (140267268077312) - DEBUG 20200522 11:12:37.183 EDT KafkaReader.C#event_cb:72] 7, facility SASLMECHS, str [thrd:GroupCoordinator]: GroupCoordinator/1: Broker supported SASL mechanisms: GSSAPI
[127201 (140267268077312) - DEBUG 20200522 11:12:37.183 EDT KafkaReader.C#event_cb:72] 7, facility AUTH, str [thrd:GroupCoordinator]: GroupCoordinator/1: Auth in state AUTH_HANDSHAKE (handshake supported)
[127201 (140267268077312) - DEBUG 20200522 11:12:37.183 EDT KafkaReader.C#event_cb:72] 7, facility STATE, str [thrd:GroupCoordinator]: GroupCoordinator: Broker changed state AUTH_HANDSHAKE -> AUTH_REQ
[127201 (140267268077312) - DEBUG 20200522 11:12:37.183 EDT KafkaReader.C#event_cb:72] 7, facility BROADCAST, str [thrd:GroupCoordinator]: Broadcasting state change
[127201 (140267268077312) - DEBUG 20200522 11:12:37.187 EDT KafkaReader.C#event_cb:72] 7, facility STATE, str [thrd:GroupCoordinator]: GroupCoordinator: Broker changed state AUTH_REQ -> UP
[127201 (140267268077312) - DEBUG 20200522 11:12:37.187 EDT KafkaReader.C#event_cb:72] 7, facility BROADCAST, str [thrd:GroupCoordinator]: Broadcasting state change
[127201 (140267268077312) - DEBUG 20200522 11:12:37.187 EDT KafkaReader.C#event_cb:72] 7, facility METADATA, str [thrd:GroupCoordinator]: Skipping metadata refresh of 19 topic(s): connected: already being requested
[127201 (140266999637760) - DEBUG 20200522 11:12:37.204 EDT KafkaReader.C#event_cb:72] 7, facility METADATA, str [thrd:main]: sasl_plaintext://mwkafka-staging-01.dr.myFirm.com:9092/3: ===== Received metadata (for 19 requested topics): broker down =====
[127201 (140266999637760) - DEBUG 20200522 11:12:37.204 EDT KafkaReader.C#event_cb:72] 7, facility METADATA, str [thrd:main]: sasl_plaintext://mwkafka-staging-01.dr.myFirm.com:9092/3: ClusterId: 52dx8qR5T1KKN8UT92peWA, ControllerId: 3
[127201 (140266999637760) - DEBUG 20200522 11:12:37.204 EDT KafkaReader.C#event_cb:72] 7, facility METADATA, str [thrd:main]: sasl_plaintext://mwkafka-staging-01.dr.myFirm.com:9092/3: 5 brokers, 19 topics
[127201 (140266999637760) - DEBUG 20200522 11:12:37.204 EDT KafkaReader.C#event_cb:72] 7, facility METADATA, str [thrd:main]: sasl_plaintext://mwkafka-staging-01.dr.myFirm.com:9092/3:   Broker #0/5: mwkafka-staging-01.tbd.myFirm.com:9092 NodeId 5
[127201 (140266999637760) - DEBUG 20200522 11:12:37.204 EDT KafkaReader.C#event_cb:72] 7, facility METADATA, str [thrd:main]: sasl_plaintext://mwkafka-staging-01.dr.myFirm.com:9092/3:   Broker #1/5: mwkafka-staging-02.dr.myFirm.com:9092 NodeId 4
[127201 (140266999637760) - DEBUG 20200522 11:12:37.204 EDT KafkaReader.C#event_cb:72] 7, facility METADATA, str [thrd:main]: sasl_plaintext://mwkafka-staging-01.dr.myFirm.com:9092/3:   Broker #2/5: mwkafka-staging-01.nyc.myFirm.com:9092 NodeId 1
[127201 (140266999637760) - DEBUG 20200522 11:12:37.204 EDT KafkaReader.C#event_cb:72] 7, facility METADATA, str [thrd:main]: sasl_plaintext://mwkafka-staging-01.dr.myFirm.com:9092/3:   Broker #3/5: mwkafka-staging-02.nyc.myFirm.com:9092 NodeId 2
[127201 (140266999637760) - DEBUG 20200522 11:12:37.204 EDT KafkaReader.C#event_cb:72] 7, facility METADATA, str [thrd:main]: sasl_plaintext://mwkafka-staging-01.dr.myFirm.com:9092/3:   Broker #4/5: mwkafka-staging-01.dr.myFirm.com:9092 NodeId 3
[127201 (140266999637760) - DEBUG 20200522 11:12:37.204 EDT KafkaReader.C#event_cb:72] 7, facility METADATA, str [thrd:main]: sasl_plaintext://mwkafka-staging-01.dr.myFirm.com:9092/3:   Topic #0/19: dips.625.20.agg with 1 partitions
[127201 (140266999637760) - DEBUG 20200522 11:12:37.204 EDT KafkaReader.C#event_cb:72] 7, facility METADATA, str [thrd:main]:   Topic dips.625.20.agg partition 0 Leader 5
```

Questions here:
1. My subscriber application did not crash here. But was crashing once a week or so. Could it be because, when the Group coordinator goes down and some messages comes in at that time, it crashes ?

2. Can I simple ignore this error as it will automatically retry ?

**Comment by edenhill** [comment]
What librdkafka version are you on?

**Comment by dshivashankar1994** [comment]
1.2.2

**Comment by edenhill** [comment]
That looks like an idle disconnect, which are fine.

If you have a crash on the latest librdkafka version, please create a new issue on librdkafka (not kafkacat) witth provided stacktrace.
